[mlir][llvmir][debug] Correctly generate location for phi nodes.

abidh opened the pull request:
In [convertBlockImpl](https://github.com/llvm/llvm-project/blob/87eeed1f0ebe57abffde560c25dd9829dc6038f3/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp#L959), the debug location is set on the builder before the op is processed. This results in correct location being given to corresponding llvm instructions. But same is not done when phi nodes are created a few lines above. This result is phi nodes getting whatever the current debug location of the builder is. It can be nothing or in worst case a stale location. Fixed by calling SetCurrentDebugLocation before generating phi nodes.

abidh committed:
[mlir][llvmir][debug] Correctly generate location for phi nodes.

In convertBlockImpl, the debug location is set on the builder before
the op is processed. This results in correct location being given to
corresponding llvm instructions. But same is not done when phi nodes
are created a few lines above. This result is phi nodes getting whatever
the current debug location of the builder is. It can be nothing or in
worst case a stale location. Fixed by calling SetCurrentDebugLocation
before generating phi nodes.

gysit commented on mlir/test/Target/LLVMIR/llvmir-phi-loc.mlir:
```diff
@@ -0,0 +1,47 @@
+// RUN: mlir-translate -mlir-to-llvmir %s | FileCheck %s
+
+
+module attributes {} {
```
```suggestion

```
nit: drop module and extra newline.

abidh replied:
Done.

gysit commented on mlir/test/Target/LLVMIR/llvmir-phi-loc.mlir:
```diff
@@ -0,0 +1,47 @@
+    %3 = llvm.mlir.constant(100 : index) : i64 loc(#loc2)
+    %7 = llvm.trunc %0 : i64 to i32 loc(#loc2)
+    llvm.br ^bb1(%7, %3 : i32, i64) loc(#loc2)
+  ^bb1(%8: i32 loc(#loc4), %9: i64 loc(#loc5)):  // 2 preds: ^bb0, ^bb2
```
Can you shorten the program a bit? It seems like two constants plus a branch to a follow up block with two arguments and a return should be sufficient for the test?

The subprogram below seems also unnecessary since we do not check if it is exported correctly. I think it is fine to omit it here.

abidh replied:
Thanks for your review. I have shortened the program. Although the subprogram bit is needed. Without a fused location on function, no location information is emitted.

gysit replied:
Ok!

gysit commented on mlir/test/Target/LLVMIR/llvmir-phi-loc.mlir:
```diff
@@ -0,0 +1,47 @@
+
+
+module attributes {} {
+  llvm.func @test(%arg0: !llvm.ptr)  {
```
```suggestion
  llvm.func @test_phi_locations(%arg0: !llvm.ptr)  {
```

abidh replied:
Done.

gysit reviewed (commented):
Thanks!

I think the test program can be shortened a bit. Otherwise LGTM.

River707 commented:
LGTM with comments resolved

abidh committed:
Handle review comments.

Remove the things from the test case which were not really essential for the test.

Dinistro commented on mlir/test/Target/LLVMIR/llvmir-phi-loc.mlir:
```diff
@@ -0,0 +1,32 @@
+#loc3 = loc("test.f90":9:5)
+#loc4 = loc(fused<#sp>[#loc1])
+
+// CHECK-LABEL: define void @test
```
Nit: This should also reflect the function name change you did

abidh replied:
Thanks for catching this. Fixed now.

Dinistro reviewed (approved):
LGTM % last nit. Thanks for fix :)

abidh committed:
Handle review comments.

Use the correct name of the function in CHECK.
